### PR TITLE
UCT/API: make uct_ep_create backward compatible

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
## Why ?
First trying to make UCT backward compatible.

## How ?
 - The same way as UCP, adding `params` argument with `field_mask` to API routines.
 - uct_ep_create, uct_ep_create_connected and uct_ep_create_sockaddr are merged into single routine

@yosefe

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8493/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8501/ for details (Mellanox internal link).

File: src/uct/api/uct.h [hunk 1]
```diff
@@ -470,6 +469,25 @@ typedef enum {
 } uct_mem_advice_t;
 
 
+/**
+ * @ingroup UCT_RESOURCE
+ * @brief UCT endpoint created by @ref uct_ep_create_sockaddr parameters field
+ *        mask.
+ *
+ * The enumeration allows specifying which fields in @ref
+ * uct_ep_sockaddr_params_t are present. It is used for the enablement of
```
Comment (brminich) [inline src/uct/api/uct.h]: minor: I'd rephrase it, like `It provides backward compatibility support`

File: src/uct/api/uct.h [hunk 2]
```diff
@@ -470,6 +469,25 @@ typedef enum {
...
+ * backward compatibility support.
+ */
+enum uct_ep_sockaddr_params_field {
+    UCT_EP_SOCKADDR_PARAM_FIELD_SOCKADDR  = UCS_BIT(0), /**< Remote sockaddr. */
```
Comment (brminich) [inline src/uct/api/uct.h]: minor: IMO, no need for `.` after a short comment

File: src/uct/api/uct_def.h [hunk 3]
```diff
@@ -405,4 +406,41 @@ typedef ucs_status_t (*uct_tag_unexp_rndv_cb_t)(void *arg, unsigned flags,
...
+typedef struct uct_ep_sockaddr_params {
+    /**
+     * Mask of valid fields in this structure, using bits from
+     * @ref uct_ep_sockaddr_params_field. Fields not specified in this mask
+     * would be ignored. Provides ABI compatibility with respect to adding new
+     * fields.
+     */
+    uint64_t                          field_mask;
+    /**
+     * The sockaddr to connect to on the remote peer. Mandatory filed for
```
Comment (brminich) [inline src/uct/api/uct_def.h]: `filed`->`field`

File: src/uct/api/uct_def.h [hunk 4]
```diff
@@ -405,4 +406,41 @@ typedef ucs_status_t (*uct_tag_unexp_rndv_cb_t)(void *arg, unsigned flags,
...
+     * a client.
+     */
+    const ucs_sock_addr_t             *sockaddr;
+    /**
+     * Interface to create the endpoint on. Mandatory filed for a client.
```
Comment (brminich) [inline src/uct/api/uct_def.h]: filed->field

Comment (brminich) [reply src/uct/api/uct_def.h]: can these params be used for a server? If no, would be enough to say that `it is mandatory field`.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8533/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8560/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8563/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8564/ for details (Mellanox internal link).

File: src/uct/api/uct_def.h [hunk 5]
```diff
@@ -405,4 +406,62 @@ typedef ucs_status_t (*uct_tag_unexp_rndv_cb_t)(void *arg, unsigned flags,
                                                 const void *rkey_buf);
 
 
+/**
+ * @ingroup UCT_RESOURCE
+ * @brief Tuning parameters for an UCT endpoint created by @ref uct_ep_create.
+ */
+typedef struct uct_ep_params {
```
Comment (yosefe) [inline src/uct/api/uct_def.h]: should be in uct.h

File: src/uct/api/uct_def.h [hunk 6]
```diff
@@ -405,4 +406,62 @@ typedef ucs_status_t (*uct_tag_unexp_rndv_cb_t)(void *arg, unsigned flags,
                                                 const void *rkey_buf);
 
 
+/**
+ * @ingroup UCT_RESOURCE
+ * @brief Tuning parameters for an UCT endpoint created by @ref uct_ep_create.
```
Comment (yosefe) [inline src/uct/api/uct_def.h]: Parameters for creating a UCT endpoint by @ref uct_ep_create
see also https://ontariotraining.net/grammar-tip-a-or-an-with-acronyms/

File: src/uct/api/uct_def.h [hunk 7]
```diff
@@ -405,4 +406,62 @@ typedef ucs_status_t (*uct_tag_unexp_rndv_cb_t)(void *arg, unsigned flags,
...
+ */
+typedef struct uct_ep_params {
+    /**
+     * Mask of valid fields in this structure, using bits from
+     * @ref uct_ep_params_field. Fields not specified in this mask
```
Comment (yosefe) [inline src/uct/api/uct_def.h]: s/in/by/

File: src/uct/api/uct_def.h [hunk 8]
```diff
@@ -405,4 +406,62 @@ typedef ucs_status_t (*uct_tag_unexp_rndv_cb_t)(void *arg, unsigned flags,
...
+     * would be ignored. Provides ABI compatibility with respect to adding new
```
Comment (yosefe) [inline src/uct/api/uct_def.h]: can remove "Provides ABI compatibility with respect to adding new.."

File: src/uct/api/uct_def.h [hunk 9]
```diff
@@ -405,4 +406,62 @@ typedef ucs_status_t (*uct_tag_unexp_rndv_cb_t)(void *arg, unsigned flags,
...
+     * fields.
+     */
+    uint64_t                          field_mask;
+    /**
```
Comment (yosefe) [inline src/uct/api/uct_def.h]: pls add space line before each comment for a new field

File: src/uct/api/uct_def.h [hunk 10]
```diff
@@ -405,4 +406,62 @@ typedef ucs_status_t (*uct_tag_unexp_rndv_cb_t)(void *arg, unsigned flags,
...
+     * The sockaddr to connect to on the remote peer. If set, @ref uct_ep_create
+     * will create an endpoint for a connection to the remote peer, specified by
+     * its socket address.
+     * @note The interface in this routine requires the
+     * @ref UCT_IFACE_FLAG_CONNECT_TO_SOCKADDR capability.
+     */
+    const ucs_sock_addr_t             *sockaddr;
+    /**
+     * Required @ref uct_cb_flags to indicate where callbacks can be invoked
+     * from.
```
Comment (yosefe) [inline src/uct/api/uct_def.h]: It's not always required, right? only for sockaddr mode?
what does it mean "from"? need to mention thread-safety

File: src/uct/api/uct_def.h [hunk 11]
```diff
@@ -405,4 +406,62 @@ typedef ucs_status_t (*uct_tag_unexp_rndv_cb_t)(void *arg, unsigned flags,
...
+     */
+    uint32_t                          sockaddr_cb_flags;
+    /**
+     * Callback for filling the user's private data, valid only if
```
Comment (yosefe) [inline src/uct/api/uct_def.h]: this is paragraph seems very confusing to anyone reading it for 1st time.. what is the private data? when is this callback called? in which case it's needed?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8596/ for details (Mellanox internal link).

File: src/uct/api/uct.h [hunk 12]
```diff
@@ -1303,27 +1385,19 @@ ucs_status_t uct_iface_reject(uct_iface_h iface,
...
- * @ingroup UCT_RESOURCE
- * @brief Create an endpoint which is connected to remote interface.
  *
- * requires @ref UCT_IFACE_FLAG_CONNECT_TO_IFACE capability.
- *
- * @param [in]  iface       Interface to create the endpoint on.
- * @param [in]  dev_addr    Remote device address to connect to.
- * @param [in]  iface_addr  Remote interface address to connect to.
- * @param [out] ep_p        Filled with handle to the new endpoint.
+ * @return UCS_OK       The endpoint is created successfully. This does not
```
Comment (yosefe) [inline src/uct/api/uct.h]: I would add to doc the 3 connection modes:
1. no address -> need to use ep_connect_to_ep later, requires iface flag ...
2. dev_addr+iface_addr -> connect to remote interface, requires iface flag ...
3. sockaddr -> connect to remote server, requires iface flag ...

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c [hunk 13]
```diff
@@ -10,24 +10,30 @@
...
-                           const uct_iface_addr_t *iface_addr)
+static UCS_CLASS_INIT_FUNC(uct_cuda_ipc_ep_t, const uct_ep_params_t *params)
 {
-    uct_cuda_ipc_iface_t *iface = ucs_derived_of(tl_iface, uct_cuda_ipc_iface_t);
+    uct_cuda_ipc_iface_t *iface = ucs_derived_of(params->iface,
+                                                 uct_cuda_ipc_iface_t);
     ucs_status_t status;
     char target_name[64];
 
+    UCT_CHECK_PARAM(ucs_test_all_flags(params->field_mask,
```
[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/uct/cuda/cuda_ipc/cuda_ipc_ep.c]: rdmacm won't work this way

File: src/uct/api/uct.h (hunk 12 above)
Comment (evgeny-leksikov) [reply src/uct/api/uct.h]: Added

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/cuda/cuda_ipc/cuda_ipc_ep.c]: so at least some ifaces it can be common macro?

[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/uct/cuda/cuda_ipc/cuda_ipc_ep.c]: Ok

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8631/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8637/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8638/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@shamisp @bbenton can you pls take a look?

**Comment by shamisp** [comment]
@tonycurtis can you please take a look at documentation as well. thanks !

**Comment by shamisp** [comment]
Adding @hjelmn

File: src/uct/api/uct_def.h (hunk 8 above)
Comment (tonycurtis) [reply src/uct/api/uct_def.h]: "would" -> "will" ?

File: src/uct/api/uct.h (hunk 12 above)
Comment (tonycurtis) [reply src/uct/api/uct.h]: field_mask: would -> will

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8647/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:bgate:retest

**Comment by evgeny-leksikov** [comment]
bot:bgate:retest

File: src/uct/api/uct.h [hunk 14]
```diff
@@ -667,6 +698,70 @@ struct uct_iface_params {
...
+     */
+    uct_iface_h                       iface;
+
+    /**
+     * User data associated with the endpoint.
+     */
+    void                              *user_data;
+
+    /**
+     * The device address to connect to on the remote peer. Should be defined
```
Comment (bbenton) [inline src/uct/api/uct.h]: s/Should be/This must be/

File: src/uct/api/uct.h [hunk 15]
```diff
@@ -667,6 +698,70 @@ struct uct_iface_params {
...
+     * together with @ref uct_ep_params_t::iface_addr to create an endpoint
+     * connected to a remote interface.
+     */
+    const uct_device_addr_t           *dev_addr;
+
+    /**
+      * Create an endpoint which is connected to remote interface.
+      * The interface address to connect to on the remote peer.
+      * @note requires @ref UCT_IFACE_FLAG_CONNECT_TO_IFACE capability.
+      */
```
Comment (bbenton) [inline src/uct/api/uct.h]: Suggested re-wording:
This specifies the remote address to use when creating an endpoint that is connected to a remote interface.
`@note`  This requires `@ref` UCT_IFACE_FLAG_CONNECT_TO_IFACE capability.

File: src/uct/api/uct.h [hunk 16]
```diff
@@ -667,6 +698,70 @@ struct uct_iface_params {
...
+     * @ref uct_cb_flags to indicate @ref uct_ep_params_t::sockaddr_pack_cb
+     * behavior. If @ref uct_ep_params_t::sockaddr_pack_cb is not set, this
+     * field will be ignored.
+     */
+    uint32_t                          sockaddr_cb_flags;
+
+    /**
+     * Callback for filling the user's private data which will be delivered to
+     * server side by @ref uct_sockaddr_conn_request_callback_t . This field is
+     * valid only if @ref uct_ep_params_t::sockaddr is set.
```
Comment (bbenton) [inline src/uct/api/uct.h]: Suggested re-wording:
Callback that will be used for filling the user's private data to be delivered to the server by `@ref` uct_sockaddr_conn_request_callback_t. This field is only valid if `@ref` uct_ep_params_t::sockaddr is set.

File: src/uct/api/uct.h [hunk 17]
```diff
@@ -1303,27 +1398,29 @@ ucs_status_t uct_iface_reject(uct_iface_h iface,
...
- * @param [in]  iface   Interface to create the endpoint on.
+ * Create a UCT endpoint in one form available modes:
+ * -# not connected if no any type of address is present, to connect the
+ *    endpoint to remote one need to call @ref uct_ep_connect_to_ep later,
+ *    requires @ref UCT_IFACE_FLAG_CONNECT_TO_EP
+ * -# connected to remote interface if @ref uct_ep_params_t::dev_addr and
+ *    @ref uct_ep_params_t::iface_addr are set, requires iface flag
+ *    @ref UCT_IFACE_FLAG_CONNECT_TO_IFACE
+ * -# connected to socket address if @ref uct_ep_params_t::sockaddr is set,
+ *    requires iface flag @ref UCT_IFACE_FLAG_CONNECT_TO_SOCKADDR
```
Comment (bbenton) [inline src/uct/api/uct.h]: Suggested re-wording.  Please ensure I have the semantics accurately captured:

Create a UCT endpoint in one of the available modes:
-# Unconnected endpoint: If no `@ref` uct_ep_params address is present, this creates an unconnected endpoint. To establish a connection to a remote endpoint, `@ref` uct_ep_connect_to_ep will need to be called.  Use of this mode requires the `@ref` UCT_IFACE_FLAG_CONNECT_TO_EP flag to be set.
-# Connect to a remote interface:  If `@ref` uct_ep_params_t::dev_addr and `@ref` uct_ep_params_t::iface_addr are set, this will establish an endpoint that is connected to a remote interface.  This requires that the `@ref` UCT_IFACE_FLAG_CONNECT_TO_IFACE flag is set.
-# Connect to a remote socket address:  If `@ref` uct_ep_params_t::sockaddr is set, this will create an endpoint that is conected to a remote socket. This requires that the `@ref` UCT_IFACE_FLAG_CONNECT_TO_SOCKADDR flag is set.

File: src/uct/api/uct.h [hunk 18]
```diff
@@ -1303,27 +1398,29 @@ ucs_status_t uct_iface_reject(uct_iface_h iface,
...
- * @param [in]  dev_addr    Remote device address to connect to.
- * @param [in]  iface_addr  Remote interface address to connect to.
- * @param [out] ep_p        Filled with handle to the new endpoint.
+ * @return UCS_OK       The endpoint is created successfully. This does not
+ *                      guarantee that the endpoint has been connected to
+ *                      the destination defined in @a params; in case of failure,
+ *                      the error will be reported to the interface error
+ *                      handler callback provided to @ref uct_iface_open
+ *                      via @ref uct_iface_params_t.err_handler.
+ * @return error code   In case of an error. (@ref ucs_status_t)
```
Comment (bbenton) [inline src/uct/api/uct.h]: Suggest the following to be more consistent with existing uct and ucp error code descriptions:
`@return` Error code as defined by `@ref` ucs_status_t

**Comment by evgeny-leksikov** [comment]
@bbenton , @tonycurtis thank you for review! your comment are addressed.
@shamisp pls take a look.

**Comment by shamisp** [comment]
I think it will break OpenMPI UCT integration code...
